pkg/controller/utils/elasticsearch.go

```diff
@@ -62,3 +103,271 @@ func GetElasticsearchClusterConfig(ctx context.Context, cli client.Client) (*ren
+		"tigera_secure_ee_dns":   buildILMPolicy(totalEsStorage, majorPctOfTotalDisk, 0.05, 8),
+		"tigera_secure_ee_bgp":   buildILMPolicy(totalEsStorage, majorPctOfTotalDisk, 0.05, 8),
+
+		"tigera_secure_ee_audit_ee":           buildILMPolicy(totalEsStorage, minorPctOfTotalDisk, pctOfDisk, int(*ls.Spec.Retention.AuditReports)),
+		"tigera_secure_ee_audit_kube":         buildILMPolicy(totalEsStorage, minorPctOfTotalDisk, pctOfDisk, int(*ls.Spec.Retention.AuditReports)),
+		"tigera_secure_ee_snapshots":          buildILMPolicy(totalEsStorage, minorPctOfTotalDisk, pctOfDisk, int(*ls.Spec.Retention.Snapshots)),
+		"tigera_secure_ee_compliance_reports": buildILMPolicy(totalEsStorage, minorPctOfTotalDisk, pctOfDisk, int(*ls.Spec.Retention.ComplianceReports)),
+		"tigera_secure_ee_benchmark_results":  buildILMPolicy(totalEsStorage, minorPctOfTotalDisk, pctOfDisk, 91),
+		"tigera_secure_ee_events":             buildILMPolicy(totalEsStorage, minorPctOfTotalDisk, pctOfDisk, 91),
+	}
+}
+
+func (es *EsClient) createOrUpdatePolicies(ctx context.Context, listPolicy map[string]policyDetail) error {
+	for indexName, pd := range listPolicy {
+		policyName := indexName + "_policy"
+
+		res, err := es.client.XPackIlmGetLifecycle().Policy(policyName).Do(ctx)
+		if err != nil {
+			if eerr, ok := err.(*elastic.Error); ok && eerr.Status == 404 {
+				// If policy doesn't exist, create one
+				return applyILMPolicy(ctx, es.client, indexName, pd.policy)
+			}
+			return err
+		}
+
+		// If policy exists, check if it needs to be updated
+		currentMaxAge, currentMaxSize, currentMinAge, err := extractPolicyDetails(res[policyName].Policy)
+		if err != nil {
+			return err
+		}
+		if currentMaxAge != pd.rolloverAge ||
+			currentMaxSize != pd.rolloverSize ||
+			currentMinAge != pd.deleteAge {
+			return applyILMPolicy(ctx, es.client, indexName, pd.policy)
+		}
+	}
+	return nil
+}
+
+func buildILMPolicy(totalEsStorage int64, totalDiskPercentage float64, percentOfDiskForLogType float64, retention int) policyDetail {
+	pd := policyDetail{}
+	pd.rolloverSize = calculateRolloverSize(totalEsStorage, totalDiskPercentage, percentOfDiskForLogType)
+	pd.rolloverAge = calculateRolloverAge(retention)
+	pd.deleteAge = fmt.Sprintf("%dd", retention)
+
+	pd.policy = map[string]interface{}{
+		"policy": map[string]interface{}{
+			"phases": map[string]interface{}{
+				"hot": map[string]interface{}{
+					"actions": map[string]interface{}{
+						"rollover": map[string]interface{}{
+							"max_size": pd.rolloverSize,
+							"max_age":  pd.rolloverAge,
+						},
+						"set_priority": map[string]interface{}{
+							"priority": 100,
+						},
+					},
+				},
+				"warm": map[string]interface{}{
+					"actions": map[string]interface{}{
+						"readonly": map[string]interface{}{},
+						"set_priority": map[string]interface{}{
+							"priority": 50,
+						},
+					},
+				},
+				"delete": map[string]interface{}{
+					"min_age": pd.deleteAge,
+					"actions": map[string]interface{}{
```

Comment:
It seems like the whole actions section can be omitted, or does es not allow that?

Reply:
correct, action is explicitly needed for ES.